### PR TITLE
New build operator integration test when pushed to main

### PR DESCRIPTION
**Pull request by bhanuba**
**Description of the issue**
We currently have EKS addon integration tests, but nothing to test new builds of the operator that runs on push to main. So we are creating a test for the operator when pushed to main.

**Description of changes**

Creating Cloudwatch Agent daemonset with checking if it is running or not.
Creating a pod with java annotation and checking if default environmental variables are injected or not.
Creating a resource of kind instrumentation and checking environment variables are injected in pod within same namespace if jave instumentation is enable.
**License**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Tests**
Integration Test: https://github.com/aws/amazon-cloudwatch-agent-operator/actions/runs/7198738382/job/19608857980?pr=50

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/workflows/OperatorIntegrtionTest.yml
Comment (lisguo) [inline .github/workflows/OperatorIntegrtionTest.yml:8]: can we change this to execute on pull request?

File: .github/workflows/OperatorIntegrtionTest.yml
Comment (lisguo) [inline .github/workflows/OperatorIntegrtionTest.yml:4]: Can we change the name of this workflow to follow the same convention as the others?

`operator-integration-test.yml`

File: .github/workflows/OperatorIntegrtionTest.yml
Comment (lisguo) [inline .github/workflows/OperatorIntegrtionTest.yml:56]: nit: You can actually use helm to apply the template for amazoncloudwatchagent by using this one-liner:

```
helm template --namespace amazon-cloudwatch -s templates/cloudwatch-agent-daemonset.yaml ./helm | kubectl apply --namespace amazon-cloudwatch -f -
```

then we don't need the amazoncloudwatchagent.yaml at all

File: .github/workflows/OperatorIntegrtionTest.yml
Comment (lisguo) [inline .github/workflows/OperatorIntegrtionTest.yml:58]: probably don't need `kubectl get pods -A` here if we are gonna get pods of the namespace the next line

File: .github/workflows/OperatorIntegrtionTest.yml
Comment (lisguo) [inline .github/workflows/OperatorIntegrtionTest.yml:59]: nit: would it be better to have a go script to check if the cloudwatch agent is running? Similar to what we did for the namespace tests.

File: .github/workflows/OperatorIntegrtionTest.yml
Comment (lisguo) [inline .github/workflows/OperatorIntegrtionTest.yml:76]: nit: would be cleaner to use absolute path here `go run cmd/validate_instrumentation_vars.go default default_instrumentation_env_variables.json`

File: .github/workflows/OperatorIntegrtionTest.yml
Comment (lisguo) [inline .github/workflows/OperatorIntegrtionTest.yml:81]: We don't need this file. We can just have the `sample-instrumentation.yaml` be created in the `default` namespace.

Then we can restart the test deployment with
```
kubectl rollout restart deployment nginx
```

**Review by lisguo — CHANGES_REQUESTED** [review]

File: .github/workflows/OperatorIntegrtionTest.yml
Comment (bhanuba) [reply .github/workflows/OperatorIntegrtionTest.yml]: the status is not always running it is erroring out as we are running on minikube.

File: .github/workflows/OperatorIntegrtionTest.yml
Comment (lisguo) [inline .github/workflows/OperatorIntegrtionTest.yml:4]: the file yaml name has to change as well

File: .github/workflows/OperatorIntegrtionTest.yml
Comment (lisguo) [inline .github/workflows/OperatorIntegrtionTest.yml:12]: nice!

File: integration-tests/operator/amazoncloudwatchagent.yaml
Comment (lisguo) [inline integration-tests/operator/amazoncloudwatchagent.yaml:1]: we can delete this file now that we are using helm templates

File: integration-tests/operator/ns-test-deployment.yaml
Comment (lisguo) [inline integration-tests/operator/ns-test-deployment.yaml:1]: we can delete this file too

File: integration-tests/operator/cmd/validate_instrumentation_vars.go
Comment (lisguo) [inline integration-tests/operator/cmd/validate_instrumentation_vars.go:29]: this if/else statement can be removed since we are passing in the namespace

so we can just do
```
args = []string{"get", "pods", "-n", namespace, "-l", "app=nginx", "-o=jsonpath='{.items[*].metadata.name}'"}
```

File: integration-tests/operator/amazoncloudwatchagent.yaml
Comment (bhanuba) [reply integration-tests/operator/amazoncloudwatchagent.yaml:1]: done!!

File: integration-tests/operator/ns-test-deployment.yaml
Comment (bhanuba) [reply integration-tests/operator/ns-test-deployment.yaml:1]: done!!

**Review by lisguo — DISMISSED** [review]

File: .github/workflows/operator-integration-test.yml
Comment (nathalapooja) [inline .github/workflows/operator-integration-test.yml:55]: I see make deploy is running the below command: 
set-image-controller
	$(KUSTOMIZE) build config/default | kubectl apply -f -

does it deploy the public ecr image on the minicube ?

File: .github/workflows/operator-integration-test.yml
Comment (nathalapooja) [inline .github/workflows/operator-integration-test.yml:51]: I see in the build run , below command was executed to docker build
docker buildx build --load --platform linux/amd64 -t public.ecr.aws/cloudwatch-agent/cloudwatch-agent-operator:1.0.2 --build-arg VERSION_PKG="github.com/aws/amazon-cloudwatch-agent-operator/internal/version" --build-arg VERSION=1.0.2 --build-arg VERSION_DATE=2023-12-09T05:03:05Z --build-arg AGENT_VERSION="1.300031.1b317" --build-arg AUTO_INSTRUMENTATION_JAVA_VERSION="v1.31.1" .

Any reason to build this to public ecr image?

File: .github/workflows/operator-integration-test.yml
Comment (bhanuba) [reply .github/workflows/operator-integration-test.yml:55]: yes we are using minikube and public ecr image gets deployed to the minikube.

File: .github/workflows/operator-integration-test.yml
Comment (sky333999) [inline .github/workflows/operator-integration-test.yml:67]: Where are we checking that it is running?

File: .github/workflows/operator-integration-test.yml
Comment (sky333999) [inline .github/workflows/operator-integration-test.yml:73]: nit: I would call the directory `manifests` instead of `operator`. And call the file `sample-deployment.yaml` instead of `test-deployment.yaml` so its extensible for future tests.

File: .github/workflows/operator-integration-test.yml
Comment (lisguo) [reply .github/workflows/operator-integration-test.yml:51]: this is just the name of the container image after we run `make container`. When we deploy to minikube, it will take preference to the local image from `make container` which is the public ecr name. 

if this is too confusing, we can change `make container` to output something like `cloudwatch-agent-operator` and it would still work.

File: Makefile
Comment (sky333999) [inline Makefile:2]: Was there a reason this had to change?

File: config/manager/kustomization.yaml
Comment (sky333999) [inline config/manager/kustomization.yaml:8]: Is this being used for `make deploy` as well? 
Did we actually push to our public repo while running these github tests? Id imagine it atleast complains about permission issues. Mind checking?

File: .github/workflows/operator-integration-test.yml
Comment (bhanuba) [reply .github/workflows/operator-integration-test.yml:51]: Here we are not actually using the public image. We are building a new image with "make container" with public ecr name.

File: config/manager/kustomization.yaml
Comment (sky333999) [reply config/manager/kustomization.yaml:8]: We shouldnt be using our public repo and tags here.

File: .github/workflows/operator-integration-test.yml
Comment (sky333999) [inline .github/workflows/operator-integration-test.yml:79]: We are introducing a new pattern here that is diff from all the other integ tests we have currently that leverage tf. Would have been nice to just stay consistent.

File: integration-tests/operator/cmd/validate_instrumentation_vars.go
Comment (sky333999) [inline integration-tests/operator/cmd/validate_instrumentation_vars.go:29]: Using the k8s go client would have been a lot cleaner. You would need to do all the unmarshalling etc. and now we have a weird mix of shell and go.

File: .github/workflows/operator-integration-test.yml
Comment (nathalapooja) [reply .github/workflows/operator-integration-test.yml:51]: I believe its better to use other name from public ecr image, it makes me assume that we are testing the public ecr image directly

File: .github/workflows/operator-integration-test.yml
Comment (bhanuba) [reply .github/workflows/operator-integration-test.yml:67]: We are running on the minikube. It will eventually fail. So we are checking with the name of the pod being created.

File: .github/workflows/operator-integration-test.yml
Comment (bhanuba) [reply .github/workflows/operator-integration-test.yml:73]: done!!

File: Makefile
Comment (bhanuba) [reply Makefile:2]: Tested by giving the static value. Will change it to as it is.

File: .github/workflows/operator-integration-test.yml
Comment (sky333999) [reply .github/workflows/operator-integration-test.yml]: Why does the agent fail? What is the error message?
If the agent isnt running, we shouldnt print that the pod is running.

File: .github/workflows/operator-integration-test.yml
Comment (lisguo) [reply .github/workflows/operator-integration-test.yml:80]: I think this test is more lightweight so that we do not have to use terraform.

This is a test that can be used on every PR that spans about 6 minutes or so to quickly test the functionality of the operator.

We have the EKS / helm / E2E test that fully tests the whole system which uses terraform.

I like having the simplicity here imo.

File: Makefile
Comment (lisguo) [reply Makefile:2]: This had to change because the latest tag would have quotes around it. So the image name ended up being `public.ecr.....:"latest"`

File: .github/workflows/operator-integration-test.yml
Comment (lisguo) [reply .github/workflows/operator-integration-test.yml:52]: that's fair, we would have to just change the name in the Makefile and kustomize.

**Review by lisguo — APPROVED** [review]

File: .github/workflows/operator-integration-test.yml
Comment (lisguo) [reply .github/workflows/operator-integration-test.yml]: It fails on minikube due to IMDS concerns. But the agent being healthy isn't really in scope for the operator integration tests IMO -- we just need to test the operator webhook functionality.

The caveat would be if we add a healthcheck in the operator...

File: .github/workflows/operator-integration-test.yml
Comment (sky333999) [reply .github/workflows/operator-integration-test.yml:80]: Doing it via terraform shouldnt change the runtime (or atleast by a significant amount). We can use the same minikube resource with terraform if we want to avoid bringing up an EKS cluster. But doing so already has comprises like being unable to verify if the agent is functional.

I think we are very shortly going to add health checks so itll be interesting how we handle that then.

File: .github/workflows/operator-integration-test.yml
Comment (sky333999) [reply .github/workflows/operator-integration-test.yml]: I think we plan on adding those health checks pretty soon.. so yeah this is bound to fail soon.

File: .github/workflows/operator-integration-test.yml
Comment (bhanuba) [reply .github/workflows/operator-integration-test.yml:52]: changed to aws/cloudwatch-agent-operator:dev

File: .github/workflows/operator-integration-test.yml
Comment (nathalapooja) [inline .github/workflows/operator-integration-test.yml:9]: Lets remove this before merging to main

File: .github/workflows/operator-integration-test.yml
Comment (nathalapooja) [reply .github/workflows/operator-integration-test.yml:56]: From previous Lisa comments I see that we are not deploying the public ECR image.

File: integration-tests/operator/cmd/validate_instrumentation_vars.go
Comment (nathalapooja) [reply integration-tests/operator/cmd/validate_instrumentation_vars.go]: I agree. We can make more similar to other integ test by using k8s client

File: Makefile
Comment (nathalapooja) [inline Makefile:2]: I assume this affects the actual image build flow for public

File: config/manager/kustomization.yaml
Comment (bhanuba) [reply config/manager/kustomization.yaml]: No we are not using this in make deploy. I have updated it now.

File: Makefile
Comment (lisguo) [reply Makefile:2]: ahh, yeah I'd prefer we leave this to what is in versions.txt ... this defines the default images used for the operator if one is not provided.

File: integration-tests/manifests/sample-instrumentation.yaml
Comment (lisguo) [inline integration-tests/manifests/sample-instrumentation.yaml:13]: not sure if you need this

File: .github/workflows/operator-integration-test.yml
Comment (lisguo) [reply .github/workflows/operator-integration-test.yml:80]: that's fair, we can use a linux ec2 for example.

this actually brings up a good point though -- is cwa only supported on AWS resources? I thought it would be able to run on native k8s?

File: .github/workflows/operator-integration-test.yml
Comment (lisguo) [inline .github/workflows/operator-integration-test.yml:57]: typo: AmazonCloudWatchAgent

File: .github/workflows/operator-integration-test.yml
Comment (lisguo) [inline .github/workflows/operator-integration-test.yml:10]: can we include `synchronize, ready_for_review` ? 

This should trigger the build in this PR

**Review by lisguo — APPROVED** [review]

File: .github/workflows/operator-integration-test.yml
Comment (sky333999) [reply .github/workflows/operator-integration-test.yml:79]: It should be able to run on native k8s as well with the right creds config setup

**Review by sky333999 — APPROVED** [review]

**Comment by sky333999** [comment]
I think there is a lot that can be cleaned up here to be consistent to our other tests, but its a non-blocking suggestion.

**Pull request merged**
